AssertionError: duplicate attachment for key

## Minimized code

I haven't been able to minimize it further 😅 

```Scala
// AE-6eb7a8f351de78d430c7ca502335eb0845bdbf3f
object Foo {
  Nil {
    class i0 { a: i2 =>
      class i1
    }
  }
}
```

## Output (click arrow to expand)
<details>

```scala
exception occurred while typechecking test.scala
exception occurred while compiling test.scala
java.lang.AssertionError: assertion failed: duplicate attachment for key dotty.tools.dotc.util.Property$Key@790a251b while compiling test.scala
Exception in thread "main" java.lang.AssertionError: assertion failed: duplicate attachment for key dotty.tools.dotc.util.Property$Key@790a251b
	at dotty.DottyPredef$.assertFail(DottyPredef.scala:17)
	at dotty.tools.dotc.util.Attachment$Container.pushAttachment(Attachment.scala:118)
	at dotty.tools.dotc.ast.Trees$Tree.pushAttachment(Trees.scala:56)
	at dotty.tools.dotc.typer.Namer.record$1(Namer.scala:489)
	at dotty.tools.dotc.typer.Namer.expand(Namer.scala:492)
	at dotty.tools.dotc.typer.Namer.index$$anonfun$1(Namer.scala:773)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.typer.Namer.index(Namer.scala:773)
	at dotty.tools.dotc.typer.Namer$ClassCompleter.completeConstructor(Namer.scala:1226)
	at dotty.tools.dotc.typer.Namer$ClassCompleter.completeInCreationContext(Namer.scala:1308)
	at dotty.tools.dotc.typer.Namer$Completer.complete(Namer.scala:851)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:259)
	at dotty.tools.dotc.core.Denotations$Denotation.completeInfo$1(Denotations.scala:188)
	at dotty.tools.dotc.core.Denotations$Denotation.info(Denotations.scala:190)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.ensureCompleted(SymDenotations.scala:445)
	at dotty.tools.dotc.typer.Typer.retrieveSym(Typer.scala:2318)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2343)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2435)
	at dotty.tools.dotc.typer.Typer.op$1(Typer.scala:2502)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2511)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2514)
	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2536)
```
</details>

--- a/tests/neg/i12640.scala
+++ b/tests/neg/i12640.scala
@@ -1,0 +1,17 @@
+package x
+
+trait CpsMonad[F[_]]:
+  def pure[A](x:A): F[A]
+  def flatMap[A,B](fa:F[A])(f: A=>F[B]): F[B]
+
+abstract sealed class CpsStream[-F[_],+T]
+
+case class Cons[F[_],T](head:T, tailFun: ()=>F[CpsStream[F,T]]) extends CpsStream[F,T]
+
+case class Empty[F[_]]() extends CpsStream[F,Nothing]
+
+def unfold[S,F[_]:CpsMonad,T](s0:S)(f:S => F[Option[(S,T)]]):F[CpsStream[F,T]] =
+      summon[CpsMonad[F]].flatMap(f(s0)){
+         case Some(s1,a) => Cons(a, () => unfold(s1,f))  // error (used to crash)
+         case None => summon[CpsMonad[F]].pure(Empty[F]())
+      }

--- a/tests/neg/i9287.scala
+++ b/tests/neg/i9287.scala
@@ -1,0 +1,7 @@
+object Foo {
+  Nil {
+    class i0 { a: i2 =>  // error
+      class i1
+    }
+  }  // error
+}

--- a/tests/pos/i12896.scala
+++ b/tests/pos/i12896.scala
@@ -1,0 +1,12 @@
+trait IO[E] {
+  def map[B](f: Any => B): IO[E] = ???
+  def flatMap[C](f: Any => IO[C]): IO[E | C] = ???
+}
+
+class Test {
+  def test: Unit = {
+    val a: IO[Nothing] = ???
+
+    val d = a.flatMap(y => a.flatMap(z => a.map(_ => z)))
+  }
+}

